[TensorExpr] Cleanup logic in the TensorExpr fuser pass.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42938 [TensorExpr] Cleanup logic in the TensorExpr fuser pass.**
* #42937 [TensorExpr] Remove redundant checks from canHandle in TE fuser.
* #42936 [TensorExpr] Wrap fuser in a class.

1. Structure the logic in a more straight-forward way: instead of magic
   tricks with node iterators in a block we now have a function that
   tries to create a fusion group starting from a given node (and pull
   everything it can into it).
2. The order in which we're pulling nodes into a fusion group is now
   more apparent.
3. The new pass structure automatically allows us to support fusion
   groups of size=1.

Differential Revision: [D23084409](https://our.internmc.facebook.com/intern/diff/D23084409)

Commit:
[TensorExpr] Cleanup logic in the TensorExpr fuser pass.

1. Structure the logic in a more straight-forward way: instead of magic
   tricks with node iterators in a block we now have a function that
   tries to create a fusion group starting from a given node (and pull
   everything it can into it).
2. The order in which we're pulling nodes into a fusion group is now
   more apparent.
3. The new pass structure automatically allows us to support fusion
   groups of size=1.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Cleanup logic in the TensorExpr fuser pass."


1. Structure the logic in a more straight-forward way: instead of magic
   tricks with node iterators in a block we now have a function that
   tries to create a fusion group starting from a given node (and pull
   everything it can into it).
2. The order in which we're pulling nodes into a fusion group is now
   more apparent.
3. The new pass structure automatically allows us to support fusion
   groups of size=1.

Differential Revision: [D23084409](https://our.internmc.facebook.com/intern/diff/D23084409)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -190,118 +190,152 @@ Node* getOrCreateTensorExprSubgraph(Node* n) {
+    // We use an ordered set here to visit them in the right order: the fusion
+    // group is closer to the end of the block and we are trying to pull later
+    // nodes first.
+    std::set<Node*, nodesComparator> queue;
```
I'm trying to think of edge cases where the queue / visited set might not behave how we want them to & get in the way of fusion... One thing is the queue might get stale/incorrect if a node gets moved around as part of  `moveBeforeTopologicallyValid`. However I think a node will only get moved if we haven't been able to merge it, so I'm not sure how much this matters. 

I guess something to think about but I can't actually think of an edge case so we should go ahead with this.

ZolotukhinM replied:
That's another very good point and I'm very happy that you also answered your concern :) I didn't think of that but after your description it indeed looks like it would still work fine. I will add a comment highlighting that.

eellison reviewed (commented):
Cool! the queue / visited set is nice. I have a couple small comments.

eellison reviewed (approved):
remaining comments are easily addressable

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -190,118 +190,152 @@ Node* getOrCreateTensorExprSubgraph(Node* n) {
+      nodes_to_merge.push_back(listconstruct);
+    }
+    for (auto n : nodes_to_merge) {
+      aliasDb_->moveBeforeTopologicallyValid(n, fusion_group);
```
That's a very good point and I'm trying to think how to properly check that without adding too much ugliness to the code.

eellison replied:
You can just check that `aliasDb_->moveBeforeTopologicallyValid(n, fusion_group);` is true, and bail if it's False

ZolotukhinM replied:
It's a bit trickier than that because if it fails for ListConstruct we also want to not include the corresponding aten::cat. Currently the only way to properly do this imho is:
1) tryMove all chained nodes next to the fusion group, bail if any of the moves failed.
2) merge them into the group

We will have a side effect of moving things that we don't fuse, but I think it's a lesser evil in this case.

ZolotukhinM replied:
I ended up splitting the loop where we move nodes to the fusion group and where we actually merge them into the group - I think it should handle the case you brought up correctly with the side effect mentioned in the previous comment.

Commit:
Update on "[TensorExpr] Cleanup logic in the TensorExpr fuser pass."


1. Structure the logic in a more straight-forward way: instead of magic
   tricks with node iterators in a block we now have a function that
   tries to create a fusion group starting from a given node (and pull
   everything it can into it).
2. The order in which we're pulling nodes into a fusion group is now
   more apparent.
3. The new pass structure automatically allows us to support fusion
   groups of size=1.

Differential Revision: [D23084409](https://our.internmc.facebook.com/intern/diff/D23084409)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Cleanup logic in the TensorExpr fuser pass."


1. Structure the logic in a more straight-forward way: instead of magic
   tricks with node iterators in a block we now have a function that
   tries to create a fusion group starting from a given node (and pull
   everything it can into it).
2. The order in which we're pulling nodes into a fusion group is now
   more apparent.
3. The new pass structure automatically allows us to support fusion
   groups of size=1.

Differential Revision: [D23084409](https://our.internmc.facebook.com/intern/diff/D23084409)

[ghstack-poisoned]
